proposals/NNNN-package-manager-product-definitions.md

```diff
@@ -0,0 +1,293 @@
+    
+    - first look for a target with the specified name in the same package
+    - second, look for a product with the specified name in any package dependency
+
+    then it should be possible for the shorthand notation to be sufficient in the vast majority of cases, with the possibility of using the more verbose form in the cases in which the shorthand form really is ambiguous.
+
+2.  Alternatively, a new optional array parameter could be added to the instantiation of a target:
+
+    ```
+    Target(name: "Foo", dependencies: ["Bar", "Baz"], externalDependencies: ["SomeProduct"])
+    ```
+    
+    If we go this route, one of the difficulties is in choosing good names from the arrays.  Also, we would still need to determine whether it would be sufficient to list strings in the `externalDependencies` parameter, or whether `(<package>, <product>)` tuples would be needed.
+
+## Detailed design
+
+### Product definitions
+
+We will add a `products` parameter to the `Package()` initializer:
+    
+```
+Package(
+    name: String,
+    pkgConfig: String? = nil,
+    providers: [SystemPackageProvider]? = nil,
+    targets: [Target] = [],
+    dependencies: [Package.Dependency] = [],
+    products: [Product]? = nil,
+    exclude: [String] = []
+)
+```
+    
+The definition of the `Product` type will be:
+   
+```
+public enum Product {
+    public class AbstractProduct {
+        public let name: String
+        public let targets: [String]
+        
+        public init(name: String, targets: [String])
+    }
+    
+    public class Executable : AbstractProduct {
+        public init(name: String, targets: [String])
+    }
+    
+    public class Library : AbstractProduct {
+        public let type: LibraryType?
+    
+        public init(name: String, type: LibraryType? = nil, targets: [String])
+    }
+    
+    public enum LibraryType {
+        case .static
+        case .dynamic
+    }
+}
+```
+
+The namespacing allows the names of the product types to be kept short.
+
+As with targets, there is no semantic significance to the order of the products in the array.
+
+### Implicit products
+
+If the `products` array is omitted, the Swift Package Manager will infer a set of implicit products based on the targets in the package.
+
+The rules for implicit products are:
+
+1.  An executable product is implicitly defined for any target that produces an executable module (as defined here: https://github.com/apple/swift-package-manager/blob/master/Documentation/Reference.md).
+
+2.  If there are any library targets, a single library product is implicitly defined for all the library targets.  The name of this product is based on the name of the package.
```

Comment:
I think it needs to include Clang modules, otherwise it would be difficult to product a library that needs to be partially implemented in C, right?